native/dlaqr5.go

```diff
@@ -0,0 +1,609 @@
+		// the near-the-diagonal part of a small bulge multi-shift QR
+		// sweep. Each 6*nbmps-2 column diagonal chunk extends from
+		// column incol to column ndcol (including both column incol and
+		// column ndcol). The following loop chases a 3*nbmps column
+		// long chain of nbmps bulges 3*nbmps-2 columns to the right.
+		// (incol may be less than ktop and ndcol may be greater than
+		// kbot indicating phantom columns from which to chase bulges
+		// before they are actually introduced or to which to chase
+		// bulges beyond column kbot.)
+		for krcol := incol; krcol <= min(incol+3*nbmps-3, kbot-2); krcol++ {
+			// Bulges number mtop to mbot are active double implicit
+			// shift bulges. There may or may not also be small 2×2
+			// bulge, if there is room. The inactive bulges (if any)
+			// must wait until the active bulges have moved down the
+			// diagonal to make room. The phantom matrix paradigm
+			// described above helps keep track.
+
+			mtop := max(0, ((ktop-1)-krcol+2)/3)
+			mbot := min(nbmps, (kbot-krcol)/3) - 1
+			m22 := mbot + 1
+			bmp22 := (mbot < nbmps-1) && (krcol+3*m22 == kbot-2)
+
+			// Generate reflections to chase the chain right one
+			// column. (The minimum value of k is ktop-1.)
+			for m := mtop; m <= mbot; m++ {
+				k := krcol + 3*m
+				if k == ktop-1 {
+					impl.Dlaqr1(3, h[ktop*ldh+ktop:], ldh,
+						sr[2*m], si[2*m], sr[2*m+1], si[2*m+1],
+						v[m*ldv:m*ldv+3])
+					alpha := v[m*ldv]
+					_, v[m*ldv] = impl.Dlarfg(3, alpha, v[m*ldv+1:m*ldv+3], 1)
+					continue
+				}
+				beta := h[(k+1)*ldh+k]
+				v[m*ldv+1] = h[(k+2)*ldh+k]
+				v[m*ldv+2] = h[(k+3)*ldh+k]
+				beta, v[m*ldv] = impl.Dlarfg(3, beta, v[m*ldv+1:m*ldv+3], 1)
+
+				// A bulge may collapse because of vigilant deflation or
+				// destructive underflow. In the underflow case, try the
+				// two-small-subdiagonals trick to try to reinflate the
+				// bulge.
+				if h[(k+3)*ldh+k] != 0 || h[(k+3)*ldh+k+1] != 0 || h[(k+3)*ldh+k+2] == 0 {
+					// Typical case: not collapsed (yet).
+					h[(k+1)*ldh+k] = beta
+					h[(k+2)*ldh+k] = 0
+					h[(k+3)*ldh+k] = 0
+					continue
+				}
+
+				// Atypical case: collapsed. Attempt to reintroduce
+				// ignoring H[k+1,k] and H[k+2,k]. If the fill
+				// resulting from the new reflector is too large,
+				// then abandon it. Otherwise, use the new one.
+				var vt [3]float64
+				impl.Dlaqr1(3, h[(k+1)*ldh+k+1:], ldh, sr[2*m],
+					si[2*m], sr[2*m+1], si[2*m+1], vt[:])
+				alpha := vt[0]
+				_, vt[0] = impl.Dlarfg(3, alpha, vt[1:3], 1)
+				refsum := vt[0] * (h[(k+1)*ldh+k] + vt[1]*h[(k+2)*ldh+k])
+
+				tmp := math.Abs(h[k*ldh+k]) + math.Abs(h[(k+1)*ldh+k+1]) + math.Abs(h[(k+2)*ldh+k+2])
+				if math.Abs(h[(k+2)*ldh+k]-refsum*vt[1])+math.Abs(refsum*vt[2]) > ulp*tmp {
+					// Starting a new bulge here would create
+					// non-negligible fill. Use the old one with
+					// trepidation.
+					h[(k+1)*ldh+k] = beta
+					h[(k+2)*ldh+k] = 0
+					h[(k+3)*ldh+k] = 0
+				} else {
```

Comment:
Up to you, but you're in a loop here, so you could continue rather than else. The equivalent indentation and the alliteration of the comment sort of makes sense the way it is.